tests/integration/relations/pgbouncer_provider/application-charm/src/charm.py

```diff
@@ -0,0 +1,199 @@
+        """Event triggered when the read/write endpoints of the database change."""
+        logger.info(f"second database endpoints have been changed to: {event.endpoints}")
+
+    # Multiple database clusters events observers.
+    def _on_cluster_database_created(self, event: DatabaseCreatedEvent) -> None:
+        """Event triggered when a database was created for this application."""
+        # Retrieve the credentials using the charm library.
+        logger.info(
+            f"cluster {event.relation.app.name} credentials: {event.username} {event.password}"
+        )
+        self.unit.status = ActiveStatus(
+            f"received database credentials for cluster {event.relation.app.name}"
+        )
+
+    def _on_cluster_endpoints_changed(self, event: DatabaseEndpointsChangedEvent) -> None:
+        """Event triggered when the read/write endpoints of the database change."""
+        logger.info(
+            f"cluster {event.relation.app.name} endpoints have been changed to: {event.endpoints}"
+        )
+
+    # Multiple database clusters events observers (for aliased clusters/relations).
+    def _on_cluster1_database_created(self, event: DatabaseCreatedEvent) -> None:
+        """Event triggered when a database was created for this application."""
+        # Retrieve the credentials using the charm library.
+        logger.info(f"cluster1 credentials: {event.username} {event.password}")
+        self.unit.status = ActiveStatus("received database credentials for cluster1")
+
+    def _on_cluster1_endpoints_changed(self, event: DatabaseEndpointsChangedEvent) -> None:
+        """Event triggered when the read/write endpoints of the database change."""
+        logger.info(f"cluster1 endpoints have been changed to: {event.endpoints}")
+
+    def _on_cluster2_database_created(self, event: DatabaseCreatedEvent) -> None:
+        """Event triggered when a database was created for this application."""
+        # Retrieve the credentials using the charm library.
+        logger.info(f"cluster2 credentials: {event.username} {event.password}")
+        self.unit.status = ActiveStatus("received database credentials for cluster2")
+
+    def _on_cluster2_endpoints_changed(self, event: DatabaseEndpointsChangedEvent) -> None:
+        """Event triggered when the read/write endpoints of the database change."""
+        logger.info(f"cluster2 endpoints have been changed to: {event.endpoints}")
+
+    def _on_run_sql_action(self, event: ActionEvent):
+        logger.error(event.params)
+        relation_id = event.params["relation-id"]
+        databag = self.first_database.fetch_relation_data()[relation_id]
+        logger.error(databag)
+
+        dbname = event.params["dbname"]
+        query = event.params["query"]
+        user = databag.get("username")
+        password = databag.get("password")
+        if event.params["readonly"]:
+            host = databag.get("read-only-endpoints").split(",")[0]
```

Comment:
I think this is optional, as both endpoint are the same (what changes is just the database name, as described on README).

Reply:
You're right, but I think it's still worth verifying that the readonly endpoints are legitimate and usable.

Reply:
I agree. You can keep them.